### PR TITLE
Unit tests for tree builder, result functions and mutations

### PR DESCRIPTION
**Pull request by VeronikaSolovei9**

### CODE REVIEW

File: hooks/hookstage/bidderrequest_mutations_test.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	tests := []struct {
+		name           string
+		bidRequest     openrtb2.BidRequest
+		impIdToBidders map[string]map[string]json.RawMessage
+		extImpPrebid   openrtb_ext.ExtImpPrebid
+		expectErr      bool
+		expectData     openrtb_ext.ExtImpPrebid
+	}{
+		{
+			name:           "One imp with 2 bidders, should be changed to one bidder",
```
Comment (bsardo) [inline hooks/hookstage/bidderrequest_mutations_test.go]: Nitpick: replace spaces with dashes in all names.

Comment (bsardo) [inline hooks/hookstage/bidderrequest_mutations_test.go]: Add test cases:
`nil-req-imp`
`empty-req-imp`
`nil-req-imp-ext`
`nil-req-imp-ext-prebid`

I think these are particularly important because there is no guarantee that these fields exist or contain data as other mutations may have run beforehand that modified the request.

Comment (bsardo) [inline hooks/hookstage/bidderrequest_mutations_test.go]: `one-req-imp-with-multiple-bidders-update-existing-bidder`

File: hooks/hookstage/bidderrequest_mutations_test.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+				"bidderA": json.RawMessage(`{"paramA": "valueA"}`),
+				"bidderB": json.RawMessage(`{"paramB": "valueB"}`),
+			}},
+			expectErr: false,
+			expectData: openrtb_ext.ExtImpPrebid{Bidder: map[string]json.RawMessage{
+				"bidderA": json.RawMessage(`{"param1": "value1"}`),
+			}},
+		},
+		{
+			name:           "One imp with 2 bidders, overwrite all bidders in imp",
```
Comment (bsardo) [inline hooks/hookstage/bidderrequest_mutations_test.go]: `one-req-imp-with-multiple-bidders-update-new-bidder`

File: hooks/hookstage/bidderrequest_mutations_test.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+				"bidderA": json.RawMessage(`{"paramA": "valueA"}`),
+				"bidderB": json.RawMessage(`{"paramB": "valueB"}`),
+			}},
+			expectErr: false,
+			expectData: openrtb_ext.ExtImpPrebid{Bidder: map[string]json.RawMessage{
+				"bidderC": json.RawMessage(`{"param1": "value1"}`),
+			}},
+		},
+		{
+			name:           "No bidders in impIdToBidders",
```
Comment (bsardo) [inline hooks/hookstage/bidderrequest_mutations_test.go]: `empty-imp-map`

Comment (bsardo) [inline hooks/hookstage/bidderrequest_mutations_test.go]: Add new test case `nil-imp-map`

File: hooks/hookstage/bidderrequest_mutations_test.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+			extImpPrebid: openrtb_ext.ExtImpPrebid{Bidder: map[string]json.RawMessage{
+				"bidderA": json.RawMessage(`{"paramA": "valueA"}`),
+			}},
+			expectErr: false,
+			expectData: openrtb_ext.ExtImpPrebid{Bidder: map[string]json.RawMessage{
+				"bidderA": json.RawMessage(`{"paramA": "valueA"}`),
+			}},
+		},
+		{
+			name:           "One imp with 1 bidder, imp not found",
```
Comment (bsardo) [inline hooks/hookstage/bidderrequest_mutations_test.go]: `one-req-imp-with-one-bidder-imp-not-found`

File: hooks/hookstage/bidderrequest_mutations_test.go [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+		name           string
+		bidRequest     openrtb2.BidRequest
+		impIdToBidders map[string]map[string]json.RawMessage
+		extImpPrebid   openrtb_ext.ExtImpPrebid
+		expectErr      bool
+		expectData     openrtb_ext.ExtImpPrebid
+	}{
+		{
+			name:           "One imp with 2 bidders, should be changed to one bidder",
+			bidRequest:     openrtb2.BidRequest{Imp: []openrtb2.Imp{{ID: "ImpA"}}},
```
Comment (bsardo) [inline hooks/hookstage/bidderrequest_mutations_test.go]: This is exactly the same in all tests. Can you remove this from the test inputs and instead create this bid request object in the test body?

File: modules/prebid/rulesengine/result_functions_test.go [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+			expectErr: false,
+		},
+		{
+			name:      "Invalid Function Name",
+			funcName:  "invalidFunction",
+			params:    json.RawMessage(`{}`),
+			expectErr: true,
+		},
+		{
+			name:      "Invalid Params",
```
Comment (bsardo) [inline modules/prebid/rulesengine/result_functions_test.go]: Rename to `invalid-exclude-bidders-params`

Add new test case `invalid-include-bidders-params`

File: modules/prebid/rulesengine/result_functions_test.go [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			_, err := NewProcessedAuctionRequestResultFunction(tt.funcName, tt.params)
+			if tt.expectErr {
+				if err == nil {
+					t.Errorf("expected error but got nil")
+				}
```
Comment (bsardo) [inline modules/prebid/rulesengine/result_functions_test.go]: `assert.Error(t, err)`

File: modules/prebid/rulesengine/result_functions_test.go [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+			} else {
+				if err != nil {
+					t.Errorf("unexpected error: %v", err)
+				}
```
Comment (bsardo) [inline modules/prebid/rulesengine/result_functions_test.go]: `assert.NoError(t, err)`

File: modules/prebid/rulesengine/result_functions_test.go [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+			name:      "Invalid Params",
+			funcName:  ExcludeBiddersName,
+			params:    json.RawMessage(`invalid-json`),
+			expectErr: true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			_, err := NewProcessedAuctionRequestResultFunction(tt.funcName, tt.params)
```
Comment (bsardo) [inline modules/prebid/rulesengine/result_functions_test.go]: We should be asserting that the proper type is being returned.

Add test case field `expectType ProcessedAuctionResultFunc`
Set it to zero value object in each test case like this: `expectType: &ExcludeBidders{},`
Add assert statement:
```
v, err : = NewProcessedAuctionRequestResultFunction(tt.funcName, tt.params)
assert.IsType(t, tt.expectType, v)
```

File: modules/prebid/rulesengine/result_functions_test.go [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+}
+
+func TestExcludeBiddersCall(t *testing.T) {
+	tests := []struct {
+		name       string
+		argBidders []string
+		req        *openrtb_ext.RequestWrapper
+	}{
+		{
+			name:       "Exclude valid bidders",
```
Comment (bsardo) [inline modules/prebid/rulesengine/result_functions_test.go]: change to `exclude-one-bidder`

File: modules/prebid/rulesengine/result_functions_test.go [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+		{
+			name:       "Exclude all bidders",
+			argBidders: []string{"bidder1", "bidder2", "bidder3"},
+			req:        mockRequestWrapperWithBidders(t, []string{"bidder1", "bidder2", "bidder3"}),
+		},
+		{
+			name:       "No bidders to exclude",
+			argBidders: []string{},
+			req:        mockRequestWrapperWithBidders(t, []string{"bidder1", "bidder2", "bidder3"}),
+		},
```
Comment (bsardo) [inline modules/prebid/rulesengine/result_functions_test.go:100]: Let's add a `nil-bidders` test case.

File: modules/prebid/rulesengine/result_functions_test.go [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			eb := &ExcludeBidders{Args: config.ResultFuncParams{Bidders: tt.argBidders}}
+			result := &hs.HookResult[hs.ProcessedAuctionRequestPayload]{
+				ChangeSet: hs.ChangeSet[hs.ProcessedAuctionRequestPayload]{},
+			}
+
+			err := eb.Call(tt.req, result, rules.ResultFunctionMeta{})
```
Comment (bsardo) [inline modules/prebid/rulesengine/result_functions_test.go:115]: Should we utilize defensive programming in the `Call` function?
```
if req == nil || result == nil {
    return error.New("some error...")
}

File: modules/prebid/rulesengine/result_functions_test.go [hunk 13]
```diff
@@ -0,0 +1,286 @@
...
+	assert.Equal(t, IncludeBiddersName, actualName, "IncludeBidders name should match expected value")
+}
+
+func TestExcludeBiddersName(t *testing.T) {
+	eb := &ExcludeBidders{}
+	actualName := eb.Name()
+	assert.Equal(t, ExcludeBiddersName, actualName, "ExcludeBidders name should match expected value")
+}
+
+func TestIncludeBiddersCall(t *testing.T) {
```
Comment (bsardo) [inline modules/prebid/rulesengine/result_functions_test.go:138]: Same comments as for `TestExcludeBiddersCall`

File: hooks/hookstage/bidderrequest_mutations_test.go [hunk 14]
```diff
@@ -0,0 +1,98 @@
+package hookstage
```
Comment (bsardo) [inline hooks/hookstage/bidderrequest_mutations_test.go]: I'm just realizing now that this mutation is actually a processed auction request mutation rather than a bidder request mutation. We should move all of the mutation logic to new files called `processedauctionrequest_mutations.go` and `processedauctionrequest_mutations_test.go`.

File: modules/prebid/rulesengine/result_functions_test.go [hunk 15]
```diff
@@ -0,0 +1,286 @@
...
+
+			assert.NoError(t, err)
+			assert.NotEmptyf(t, result.ChangeSet, "change set is empty")
+			assert.Len(t, result.ChangeSet.Mutations(), 1)
+			assert.Equal(t, hs.MutationUpdate, result.ChangeSet.Mutations()[0].Type())
+		})
+	}
+}
+
+func TestBuildIncludeBidders(t *testing.T) {
```
Comment (bsardo) [inline modules/prebid/rulesengine/result_functions_test.go:183]: I think we need more test cases:
`req-is-nil`
`req-imp-is-nil`
`req-imp-is-empty`
`req-imp-ext-is-nil`
`req-imp-ext-error`
`req-imp-ext-prebid-is-nil`
`arg-bidders-is-nil-include-no-additional-bidders-in-req`
`arg-bidders-is-empty-include-no-additional-bidders-in-req`
`include-one-bidder-already-in-req`
`include-one-bidder-not-in-req`
`include-multiple-bidders-not-in-req`
`include-one-bidder-in-req-and-one-not-in-req`
`multiple-imps`

File: modules/prebid/rulesengine/result_functions_test.go [hunk 16]
```diff
@@ -0,0 +1,286 @@
...
+		t.Run(tt.name, func(t *testing.T) {
+			result, err := buildIncludeBidders(tt.req, tt.argBidders)
+			if tt.expectErr {
+				if err == nil {
+					t.Errorf("expected error but got nil")
+				}
+			} else {
+				if err != nil {
+					t.Errorf("unexpected error: %v", err)
+				}
```
Comment (bsardo) [inline modules/prebid/rulesengine/result_functions_test.go]: `assert.NoError(t, err)`

File: modules/prebid/rulesengine/result_functions_test.go [hunk 17]
```diff
@@ -0,0 +1,286 @@
...
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			result, err := buildIncludeBidders(tt.req, tt.argBidders)
+			if tt.expectErr {
+				if err == nil {
+					t.Errorf("expected error but got nil")
+				}
```
Comment (bsardo) [inline modules/prebid/rulesengine/result_functions_test.go]: `assert.Error(t, err)`

File: hooks/hookstage/bidderrequest_mutations_test.go (hunk 1 above)
Comment (VeronikaSolovei9) [reply hooks/hookstage/bidderrequest_mutations_test.go]: `nil-req-imp-ext` - cannot set up due to request wrapper always creates ImpExt even if it is nil: 
```golang
func (w *ImpWrapper) GetImpExt() (*ImpExt, error) {
	if w.impExt != nil {
		return w.impExt, nil
	}
	w.impExt = &ImpExt{}.                 //----- this line
	if w.Imp == nil || w.Ext == nil {
		return w.impExt, w.impExt.unmarshal(json.RawMessage{})
	}
	return w.impExt, w.impExt.unmarshal(w.Ext)
}
```

Comment (VeronikaSolovei9) [reply hooks/hookstage/bidderrequest_mutations_test.go]: Added other test cases.

File: modules/prebid/rulesengine/result_functions_test.go (hunk 8 above)
Comment (VeronikaSolovei9) [reply modules/prebid/rulesengine/result_functions_test.go]: I replaces the whole error check to this:
```golang 
if tt.expectErr {
	assert.Error(t, err, "expected error but got nil")
}
```

File: modules/prebid/rulesengine/result_functions_test.go (hunk 12 above)
Comment (VeronikaSolovei9) [reply modules/prebid/rulesengine/result_functions_test.go:115]: We control these parameters in the unit test itself, so they are hardcoded: 
```golang
result := &hs.HookResult[hs.ProcessedAuctionRequestPayload]{
	ChangeSet: hs.ChangeSet[hs.ProcessedAuctionRequestPayload]{},
}
```golang
and 
```
 mockRequestWrapperWithBidders(t, []string{"bidder1", "bidder2", "bidder3"}),
```
I don't think we should add the nil checks

File: modules/prebid/rulesengine/result_functions_test.go (hunk 15 above)
Comment (bsardo) [reply modules/prebid/rulesengine/result_functions_test.go:183]: Discussed offline. Adding top-level `req is nil` check so don't need the test case here. It is unclear if `req.imp` is guaranteed to be present. We will figure that out in a later phase. Skip the `arg-bidders-is-nil...` and `arg-bidders-is-empty...` cases but add the validation logic to the constructor functions to ensure the bidder lists are not empty or nil.

File: modules/prebid/rulesengine/tree_builder_test.go [hunk 18]
```diff
@@ -12,10 +12,101 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
-func TestBuildTreeFullConfig(t *testing.T) {
+func TestBuildTree(t *testing.T) {
```
Comment (bsardo) [inline modules/prebid/rulesengine/tree_builder_test.go:15]: Nitpick in `tree_builder.go`:
We can simplify this code block:
```
if ok {
	currNode = currNode.Children[condition]
} else {
	nextNode := &rules.Node[T1, T2]{}
	currNode.Children[condition] = nextNode
	currNode = nextNode
}
```

to:
```
if !ok {
	currNode.Children[condition] = &rules.Node[T1, T2]{}
}
currNode = currNode.Children[condition]
```

Comment (bsardo) [inline modules/prebid/rulesengine/tree_builder_test.go:15]: The `Build` function assumes the config is valid as it assumes that the number of schema functions matches the number of conditions. This assumption is ok but should be documented in the `Build` comment above the function signature.
Without this assumption `tb.SchemaFuncFactory(tb.Config.Schema[ci].Func, tb.Config.Schema[ci].Args)` could result in a panic as we might be attempting to a schema function that does not exist at that index. 
We are currently performing this validation in `modules/prebid/rulesengine/config/config.go#validateRuleSet`.

Given this assumption, I am good with the level of test coverage for `Build`.

File: modules/prebid/rulesengine/result_functions_test.go (hunk 15 above)
Comment (VeronikaSolovei9) [reply modules/prebid/rulesengine/result_functions_test.go:183]: Added for both Exclude and include bidders functions

File: modules/prebid/rulesengine/tree_builder_test.go (hunk 18 above)
Comment (VeronikaSolovei9) [reply modules/prebid/rulesengine/tree_builder_test.go:15]: Indeed!

File: hooks/hookstage/processedauctionrequest_mutations.go [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+				if impBidders, ok := impIdToBidders[impWrapper.ID]; ok {
+					impExt, impExtErr := impWrapper.GetImpExt()
+					if err != nil {
+						return p, impExtErr
+					}
+					impPrebid := impExt.GetPrebid()
+					if impPrebid != nil {
+						impPrebid.Bidder = impBidders
+						impExt.SetPrebid(impPrebid)
+					}
```
Comment (bsardo) [inline hooks/hookstage/processedauctionrequest_mutations.go]: I think we should update this logic so that prebid is created if it doesn't exist:
```
impPrebid := impExt.GetPrebid()
if impPrebid == nil {
	impPrebid = &openrtb_ext.ExtImpPrebid{}
}
impPrebid.Bidder = impBidders
impExt.SetPrebid(impPrebid)
```

You can then update the test `"nil-req-imp-ext-prebid"` as follows:
```
{
	name:           "nil-req-imp-ext-prebid",
	impIdToBidders: map[string]map[string]json.RawMessage{"ImpA": {"bidderA": json.RawMessage(`{"param1": "value1"}`)}},
	extImpPrebid:   nil,
	expectErr:      false,
	expectData:     openrtb_ext.ExtImpPrebid{Bidder: map[string]json.RawMessage{
		"bidderA": json.RawMessage(`{"param1": "value1"}`),
	}},
},
```

And you can delete all references to `expectEmptyImpPrebid` in `TestUpdatePrebidBidders`.

File: modules/prebid/rulesengine/result_functions_test.go [hunk 20]
```diff
@@ -0,0 +1,534 @@
...
+			argBidders: []string{"bidder4"},
+			req:        mockRequestWrapperWithEmptyImp(t),
+			expected:   map[string]map[string]json.RawMessage{},
+			expectErr:  false,
+		},
+		{
+			name:       "req-imp-ext-is-nil",
+			argBidders: []string{"bidder4"},
+			req:        mockRequestWrapperWithImpExtNil(t),
+			expectErr:  true,
```
Comment (bsardo) [inline modules/prebid/rulesengine/result_functions_test.go:228]: Why does this test case result in an error?

Comment (VeronikaSolovei9) [reply modules/prebid/rulesengine/result_functions_test.go:228]: The execution path is exactly the same as in "req-imp-ext-prebid-is-nil". Error comes from nil probid.

Comment (bsardo) [reply modules/prebid/rulesengine/result_functions_test.go:228]: Discussed offline. This behavior is ok but might be revisited at some point.

File: hooks/hookstage/processedauctionrequest_mutations.go [hunk 21]
```diff
@@ -0,0 +1,56 @@
+package hookstage
+
+import (
+	"encoding/json"
+	"errors"
+	"github.com/prebid/prebid-server/v3/openrtb_ext"
+)
+
+func (c *ChangeSet[T]) ProcessedAuctionRequest() ChangeSetProcessedAuctionRequest[T] {
```
Comment (guscarreon) [inline hooks/hookstage/processedauctionrequest_mutations.go:9]: Nitpick: Can we move `ProcessedAuctionRequest()` to where `ChangeSet[T]` is defined?

File: hooks/hookstage/processedauctionrequest_mutations.go [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+
+type ChangeSetProcessedAuctionRequest[T any] struct {
+	changeSet *ChangeSet[T]
+}
+
+func (c ChangeSetProcessedAuctionRequest[T]) Bidders() ChangeBidders[T] {
+	return ChangeBidders[T]{changeSetProcessedAuctionRequest: c}
+}
+
+type ChangeBidders[T any] struct {
```
Comment (guscarreon) [inline hooks/hookstage/processedauctionrequest_mutations.go:21]: Nitpick: can we move `ChangeBidders[T any]` right before the definition of its `Update(...)` method?

File: hooks/hookstage/processedauctionrequest_mutations.go (hunk 21 above)
Comment (guscarreon) [reply hooks/hookstage/processedauctionrequest_mutations.go:9]: I just noticed there's already an interface called `ProcessedAuctionRequest` inside `hooks/hookstage/processedauctionrequest.go`. This might lead to confusion, can we modify?

**Review by guscarreon — APPROVED** [review]
LGTM
